### PR TITLE
RSDK-8819: Add FTDC file deletion.

### PR DESCRIPTION
**Pull request by dgottlieb**

### CODE REVIEW

File: ftdc/ftdc.go [hunk 1]
```diff
@@ -439,6 +459,107 @@ func (ftdc *FTDC) getWriter() (io.Writer, error) {
...
+	}
+
+	if len(files) <= ftdc.maxNumFiles {
+		// We have yet to hit our file limit. Keep all of the files.
+		ftdc.logger.Debugw("Inside the budget for ftdc files", "numFiles", len(files), "maxNumFiles", ftdc.maxNumFiles)
+		return nil
+	}
+
+	slices.SortFunc(files, func(left, right fileTime) int {
+		// Sort in descening order. Such that files indexed first are safe. This eases walking the
```
Comment (benjirewis) [inline ftdc/ftdc.go:518]:
```suggestion
		// Sort in descending order. Such that files indexed first are safe. This eases walking the
```

File: ftdc/ftdc.go [hunk 2]
```diff
@@ -439,6 +459,107 @@ func (ftdc *FTDC) getWriter() (io.Writer, error) {
...
+			ftdc.logger.Warnw("Error removing FTDC file", "filename", file.name)
+		}
+	}
+
+	return nil
+}
+
+// filenameTimeRe matches the files produced by ftdc. Filename <-> regex parity is exercised by file
+// deletion testing. Filename generation uses padding such that we can rely on there before 2/4
+// digits for every numeric value.
```
Comment (benjirewis) [inline ftdc/ftdc.go:537]: Could you put an example of what this regex matches in this comment?

File: ftdc/ftdc_test.go [hunk 3]
```diff
@@ -400,3 +401,128 @@ func TestCountingBytes(t *testing.T) {
...
+	test.That(t, timeVal.Month(), test.ShouldEqual, time.Month(11))
+	test.That(t, timeVal.Day(), test.ShouldEqual, 18)
+	test.That(t, timeVal.Hour(), test.ShouldEqual, 20)
+	test.That(t, timeVal.Minute(), test.ShouldEqual, 37)
+	test.That(t, timeVal.Second(), test.ShouldEqual, 1)
+}
+
+func TestFileDeletion(t *testing.T) {
+	// This test takes ~10 seconds due to file naming limitations. This test creates FTDC files on
+	// disk whos names include the timestamp with seconds resolution. In this case FTDC has to wait
```
Comment (benjirewis) [inline ftdc/ftdc_test.go:418]:
```suggestion
	// disk whose names include the timestamp with seconds resolution. In this case FTDC has to wait
```

File: ftdc/ftdc_test.go [hunk 4]
```diff
@@ -400,3 +401,128 @@ func TestCountingBytes(t *testing.T) {
...
+	}
+
+	// And assert the "oldest original" files are no longer found.
+	for idx := len(leftoverFiles); idx < len(origFiles); idx++ {
+		// The `fs.FileInfo` returned by `os.Lstat` does not include the directory as part of its
+		// file name. Reconstitute the relative path before testing.
+		_, err := os.Lstat(filepath.Join(ftdc.ftdcDir, origFiles[idx].Name()))
+		var pathErr *fs.PathError
+		if !errors.As(err, &pathErr) {
+			t.Fatalf("File should be deleted. Lstat error? %v", err)
```
Comment (benjirewis) [inline ftdc/ftdc_test.go:504]:
```suggestion
			t.Fatalf("File should be deleted. Lstat error: %v", err)
```

File: ftdc/ftdc.go [hunk 5]
```diff
@@ -183,6 +191,14 @@ func (ftdc *FTDC) Remove(name string) {
 func (ftdc *FTDC) Start() {
 	ftdc.readStatsWorker = utils.NewStoppableWorkerWithTicker(time.Second, ftdc.statsReader)
 	utils.PanicCapturingGo(ftdc.statsWriter)
+
+	// The `fileDeleter` goroutine mostly aligns with the "stoppable worker with ticker"
+	// pattern. But it has the additional desire that if file deletion exits with a panic, all of
+	// FTDC should stop.
+	utils.PanicCapturingGoWithCallback(ftdc.fileDeleter, func(err any) {
```
Comment (benjirewis) [inline ftdc/ftdc.go:198]: When would `fileDeleter` panic?

**Review by benjirewis — APPROVED** [review]
Nice! Just some typo-corrections and a question.

File: ftdc/ftdc.go (hunk 5 above)
Comment (dgottlieb) [reply ftdc/ftdc.go:198]: nobody knows!

File: ftdc/ftdc.go (hunk 1 above)
Comment (dgottlieb) [reply ftdc/ftdc.go]: added locally

File: ftdc/ftdc.go (hunk 2 above)
Comment (dgottlieb) [reply ftdc/ftdc.go:537]: done

File: ftdc/ftdc_test.go (hunk 4 above)
Comment (dgottlieb) [reply ftdc/ftdc_test.go]: added locally

File: ftdc/ftdc.go (hunk 5 above)
Comment (benjirewis) [reply ftdc/ftdc.go:198]: If we aren't aware of a `panic` path, then why introduce this code? The new path to `StopAndJoin` seems almost as worrying to me as FTDC causing a `viam-server` crash.

Comment (dgottlieb) [reply ftdc/ftdc.go:198]: We use them in all kinds of places that don't actually panic. I'm torn on that decision. But most relevant, the stoppable workers is in a panic capturing go and the stats writer are in one. So if this were not in one, that would raise more suspicion. Maybe?

Comment (benjirewis) [reply ftdc/ftdc.go:198]: Discussed offline: what you have sounds good. The precedent does seem to be panic capturing.

**Pull request merged**